File: python/leetcode_grind/day_582.py
# https://leetcode.com/problems/minimum-number-of-days-to-make-m-bouquets/description
class Solution1:
    def minDays(self, bloomDay: List[int], m: int, k: int) -> int:
        def get_num_of_bouqets(bloomDay, mid, k):
            num_of_bouqets = 0
            count = 0
            for day in bloomDay:
                if day <= mid:
                    count += 1
                else:
                    count = 0
                if count == k:
                    num_of_bouqets += 1
                    count = 0
            return num_of_bouqets
        
        if m * k > len(bloomDay):
            return -1
        start = 0
        end = max(bloomDay)
        minDays = -1

        while start <= end:
            mid = (start + end) // 2
            if get_num_of_bouqets(bloomDay, mid, k) >= m:
                minDays = mid
                end = mid - 1
            else:
                start = mid + 1
        return minDays

File: java/leetcode_grind/Day582.java
package leetcode_grind;

public class Day582 {
    // https://leetcode.com/problems/minimum-number-of-days-to-make-m-bouquets/description
    static class Solution1 {
        public int minDays(int[] bloomDay, int m, int k) {
            var check = new Object() {
                boolean apply(int d) {
                    var b = 0;
                    outer: for (int i = 0; i < bloomDay.length; i++) {
                        for (int j = 0; j < k; j++) {
                            if (i + j >= bloomDay.length || bloomDay[i + j] > d) {
                                continue outer;
                            }
                        }
                        b++;
                        if (b == m) {
                            return true;
                        }
                        i += k - 1;
                    }
                    return b >= m;
                }
            };

            int n = bloomDay.length;
            if (n / k < m) {
                return -1;
            }

            int lo = -1;
            int hi = (int) 1e9;

            while (lo + 1 < hi) {
                int mid = lo + (hi - lo) / 2;
                if (check.apply(mid)) {
                    hi = mid;
                } else {
                    lo = mid;
                }
            }
            return hi == 0 ? -1 : hi;
        }
    }
}
